File: src/main/java/br/com/correios/enderecador/telas/TelaExportarDados.java
package br.com.correios.enderecador.telas;

import br.com.correios.enderecador.bean.DestinatarioBean;
import br.com.correios.enderecador.dao.DaoException;
import br.com.correios.enderecador.dao.DestinatarioDao;
import br.com.correios.enderecador.excecao.EnderecadorExcecao;
import br.com.correios.enderecador.util.EnderecadorObservable;
import br.com.correios.enderecador.util.FiltroArquivo;
import br.com.correios.enderecador.util.Geral;

import java.awt.BorderLayout;
import java.awt.Cursor;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.Frame;
import java.awt.GridLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Observable;
import java.util.Observer;
import java.util.Vector;
import javax.swing.*;

import org.apache.log4j.Logger;

public class TelaExportarDados extends JFrame implements Observer {
    private static final Logger logger = Logger.getLogger(TelaExportarDados.class);

    private final Frame frmParent;

    private final Vector<DestinatarioBean> vecDestinatario;

    private static TelaExportarDados instance;

    private JButton jbntDestinatario;

    private JButton jbntGrupo;

    private JCheckBox jchkExportarTodos;

    private JList<DestinatarioBean> jlstDestinatarios;

    private TelaExportarDados(Frame parent) {
        this.vecDestinatario = new Vector<>();
        EnderecadorObservable observable = EnderecadorObservable.getInstance();
        this.frmParent = parent;
        initComponents();
        observable.addObserver(this);
    }

    public static TelaExportarDados getInstance(Frame parent) {
        if (instance == null)
            instance = new TelaExportarDados(parent);
        return instance;
    }


    public void carregaListaDestinatario() {
        try {
            ArrayList<DestinatarioBean> arrayDestinatario = DestinatarioDao.getInstance().recuperaDestinatario("");
            this.vecDestinatario.addAll(arrayDestinatario);
            this.jlstDestinatarios.setListData(this.vecDestinatario);
        } catch (DaoException e) {
            logger.error(e.getMessage(), e);
            JOptionPane.showMessageDialog(this, "Não foi possivel carregar relação de destinatários", "Endereçador ", JOptionPane.WARNING_MESSAGE);
        }
    }

    private void initComponents() {
        JToolBar jToolBar1 = new JToolBar();
        JButton jbtExportar = new JButton();
        this.jbntDestinatario = new JButton();
        this.jbntGrupo = new JButton();
        JButton jbntExcluir = new JButton();
        JPanel jPanel1 = new JPanel();
        JPanel jPanel2 = new JPanel();
        this.jchkExportarTodos = new JCheckBox();
        JScrollPane jScrollPane1 = new JScrollPane();
        this.jlstDestinatarios = new JList<>();
        setResizable(true);
        setTitle("Exportar Dados");
        setPreferredSize(new Dimension(744, 434));
        jToolBar1.setPreferredSize(new Dimension(325, 59));
        jbtExportar.setFont(new Font("MS Sans Serif", 0, 9));
        jbtExportar.setIcon(new ImageIcon(getClass().getResource("/imagens/OK.gif")));
        jbtExportar.setText("Exportar");
        jbtExportar.setHorizontalTextPosition(0);
        jbtExportar.setMaximumSize(new Dimension(90, 60));
        jbtExportar.setVerticalTextPosition(3);
        jbtExportar.addActionListener(TelaExportarDados.this::jbtExportarActionPerformed);
        jToolBar1.add(jbtExportar);
        this.jbntDestinatario.setFont(new Font("MS Sans Serif", 0, 9));
        this.jbntDestinatario.setIcon(new ImageIcon(getClass().getResource("/imagens/addusuario.gif")));
        this.jbntDestinatario.setText("Selecionar destinatário");
        this.jbntDestinatario.setHorizontalTextPosition(0);
        this.jbntDestinatario.setMaximumSize(new Dimension(107, 60));
        this.jbntDestinatario.setVerticalTextPosition(3);
        this.jbntDestinatario.addActionListener(TelaExportarDados.this::jbntDestinatarioActionPerformed);
        jToolBar1.add(this.jbntDestinatario);
        this.jbntGrupo.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 9));
        this.jbntGrupo.setIcon(new ImageIcon(getClass().getResource("/imagens/addusuarios.gif")));
        this.jbntGrupo.setText("Selecionar grupo");
        this.jbntGrupo.setHorizontalTextPosition(0);
        this.jbntGrupo.setMaximumSize(new Dimension(90, 60));
        this.jbntGrupo.setPreferredSize(new Dimension(90, 60));
        this.jbntGrupo.setVerticalTextPosition(3);
        this.jbntGrupo.addActionListener(TelaExportarDados.this::jbntGrupoActionPerformed);
        jToolBar1.add(this.jbntGrupo);
        jbntExcluir.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 9));
        jbntExcluir.setIcon(new ImageIcon(getClass().getResource("/imagens/removerTodos.gif")));
        jbntExcluir.setText("Remover todos");
        jbntExcluir.setHorizontalTextPosition(0);
        jbntExcluir.setMaximumSize(new Dimension(90, 60));
        jbntExcluir.setVerticalTextPosition(3);
        jbntExcluir.addActionListener(TelaExportarDados.this::jbntExcluirActionPerformed);
        jToolBar1.add(jbntExcluir);
        getContentPane().add(jToolBar1, "North");
        jPanel1.setLayout(new BorderLayout());
        jPanel2.setLayout(new GridLayout(1, 0));
        jPanel2.setBorder(BorderFactory.createEtchedBorder());
        this.jchkExportarTodos.setFont(new Font("MS Sans Serif", 0, 10));
        this.jchkExportarTodos.setText("Exportar todos os destinatários");
        this.jchkExportarTodos.addActionListener(TelaExportarDados.this::jchkExportarTodosActionPerformed);
        jPanel2.add(this.jchkExportarTodos);
        jPanel1.add(jPanel2, "North");
        jScrollPane1.setViewportView(this.jlstDestinatarios);
        jPanel1.add(jScrollPane1, "Center");
        getContentPane().add(jPanel1, "Center");
        pack();
    }

    private void jbntExcluirActionPerformed(ActionEvent evt) {
        this.vecDestinatario.removeAllElements();
        this.jlstDestinatarios.setListData(this.vecDestinatario);
        this.jchkExportarTodos.setSelected(false);
        this.jbntDestinatario.setEnabled(true);
        this.jbntGrupo.setEnabled(true);
    }

    private void jbntGrupoActionPerformed(ActionEvent evt) {
        setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
        TelaPesquisarGrupo telaPesquisarGrupo = new TelaPesquisarGrupo(this.frmParent, true, this.vecDestinatario);
        telaPesquisarGrupo.setVisible(true);
        this.jlstDestinatarios.setListData(this.vecDestinatario);
        setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
    }

    private void jbntDestinatarioActionPerformed(ActionEvent evt) {
        setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
        TelaPesquisarDestinatario telaPesquisaDestinatario = new TelaPesquisarDestinatario(this.frmParent, true, this.vecDestinatario);
        telaPesquisaDestinatario.setVisible(true);
        this.jlstDestinatarios.setListData(this.vecDestinatario);
        setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
    }

    private void jchkExportarTodosActionPerformed(ActionEvent evt) {
        setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
        this.jbntDestinatario.setEnabled(!this.jbntDestinatario.isEnabled());
        this.jbntGrupo.setEnabled(!this.jbntGrupo.isEnabled());
        this.vecDestinatario.removeAllElements();
        this.jlstDestinatarios.setListData(this.vecDestinatario);
        if (this.jchkExportarTodos.isSelected())
            carregaListaDestinatario();
        setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR));
    }

    private void jbtExportarActionPerformed(ActionEvent evt) {
        Geral geral = new Geral();
        String vNomeArquivo;
        if (this.vecDestinatario.size() <= 0) {
            JOptionPane.showMessageDialog(this, "Não existe nenhum destinatário selecionado!", "Endereçador", JOptionPane.WARNING_MESSAGE);
            return;
        }
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setApproveButtonText("Salvar");
        fileChooser.setDialogTitle("Salvar arquivo");
        fileChooser.setFileSelectionMode(0);
        fileChooser.setAcceptAllFileFilterUsed(true);
        fileChooser.addChoosableFileFilter(new FiltroArquivo("csv"));
        fileChooser.addChoosableFileFilter(new FiltroArquivo("txt"));
        int result = fileChooser.showSaveDialog(this);
        if (result == 1)
            return;
        File fileName = fileChooser.getSelectedFile();
        if (fileName == null)
            return;
        String extension = Geral.getExtension(fileName);
        FiltroArquivo imageFilter = (FiltroArquivo) fileChooser.getFileFilter();
        vNomeArquivo = fileName.getAbsolutePath();
        if (extension == null) {
            vNomeArquivo = vNomeArquivo + "." + imageFilter.getExtension();
            extension = imageFilter.getExtension();
        } else if (!Geral.validaArquivo(fileName)) {
            JOptionPane.showMessageDialog(this, "Formato do arquivo inválido!\nO arquivo deve ser do tipo TXT ou CSV.", "Endereçador", JOptionPane.WARNING_MESSAGE);
            return;
        }
        try {
            if (extension.equals("txt")) {
                geral.exportarParaTXT(vNomeArquivo, this.vecDestinatario);
            } else {
                geral.exportarParaCSV(vNomeArquivo, this.vecDestinatario);
            }
            JOptionPane.showMessageDialog(this, "Exportação concluída!", "Endereçador", JOptionPane.WARNING_MESSAGE);
        } catch (EnderecadorExcecao ex) {
            JOptionPane.showMessageDialog(this, ex.getMessage(), "Endereçador", JOptionPane.WARNING_MESSAGE);
        }
    }

    public void update(Observable o, Object arg) {
        if (arg instanceof DestinatarioBean) {
            DestinatarioBean destinatario = (DestinatarioBean) arg;
            int index = this.vecDestinatario.indexOf(destinatario);
            if (index != -1) {
                this.vecDestinatario.remove(index);
                this.vecDestinatario.add(index, destinatario);
                this.jlstDestinatarios.setListData(this.vecDestinatario);
            }
        } else if (arg instanceof List) {
            List<DestinatarioBean> listaDestinatarios = (List<DestinatarioBean>) arg;
            for (DestinatarioBean listaDestinatario : listaDestinatarios) {
                if (listaDestinatario != null)
                    this.vecDestinatario.remove(listaDestinatario);
            }
            this.jlstDestinatarios.setListData(this.vecDestinatario);
        }
    }
}

File: src/main/kotlin/br/com/correios/enderecador/telas/TelaExportarDados.kt
package br.com.correios.enderecador.telas

import javax.swing.JFrame
import br.com.correios.enderecador.bean.DestinatarioBean
import javax.swing.JButton
import javax.swing.JCheckBox
import javax.swing.JList
import br.com.correios.enderecador.util.EnderecadorObservable
import br.com.correios.enderecador.dao.DestinatarioDao
import br.com.correios.enderecador.dao.DaoException
import javax.swing.JOptionPane
import javax.swing.JToolBar
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.ImageIcon
import javax.swing.BorderFactory
import br.com.correios.enderecador.util.Geral
import javax.swing.JFileChooser
import br.com.correios.enderecador.util.FiltroArquivo
import br.com.correios.enderecador.excecao.EnderecadorExcecao
import org.apache.log4j.Logger
import org.koin.core.annotation.Singleton
import java.awt.*
import java.util.*

@Singleton
class TelaExportarDados : JFrame(), Observer {
    private val vecDestinatario: Vector<DestinatarioBean?> = Vector()
    private var jbntDestinatario: JButton? = null
    private var jbntGrupo: JButton? = null
    private var jchkExportarTodos: JCheckBox? = null
    private var jlstDestinatarios: JList<DestinatarioBean>? = null

    init {
        val observable = EnderecadorObservable.instance
        initComponents()
        observable?.addObserver(this)
        setLocationRelativeTo(null)
    }

    fun carregaListaDestinatario() {
        try {
            val arrayDestinatario = DestinatarioDao.instance!!.recuperaDestinatario("")
            vecDestinatario.addAll(arrayDestinatario)
            jlstDestinatarios!!.setListData(vecDestinatario)
        } catch (e: DaoException) {
            logger.error(e.message, e)
            JOptionPane.showMessageDialog(
                this,
                "Não foi possivel carregar relação de destinatários",
                "Endereçador ",
                JOptionPane.WARNING_MESSAGE
            )
        }
    }

    private fun initComponents() {
        val jToolBar1 = JToolBar()
        val jbtExportar = JButton()
        jbntDestinatario = JButton()
        jbntGrupo = JButton()
        val jbntExcluir = JButton()
        val jPanel1 = JPanel()
        val jPanel2 = JPanel()
        jchkExportarTodos = JCheckBox()
        val jScrollPane1 = JScrollPane()
        jlstDestinatarios = JList()
        isResizable = true
        title = "Exportar Dados"
        preferredSize = Dimension(744, 434)
        jToolBar1.preferredSize = Dimension(325, 59)
        jbtExportar.font = Font("MS Sans Serif", 0, 9)
        jbtExportar.icon = ImageIcon(javaClass.getResource("/imagens/OK.gif"))
        jbtExportar.text = "Exportar"
        jbtExportar.horizontalTextPosition = 0
        jbtExportar.maximumSize = Dimension(90, 60)
        jbtExportar.verticalTextPosition = 3
        jbtExportar.addActionListener { jbtExportarActionPerformed() }
        jToolBar1.add(jbtExportar)
        jbntDestinatario!!.font = Font("MS Sans Serif", 0, 9)
        jbntDestinatario!!.icon = ImageIcon(javaClass.getResource("/imagens/addusuario.gif"))
        jbntDestinatario!!.text = "Selecionar destinatário"
        jbntDestinatario!!.horizontalTextPosition = 0
        jbntDestinatario!!.maximumSize = Dimension(107, 60)
        jbntDestinatario!!.verticalTextPosition = 3
        jbntDestinatario!!.addActionListener { jbntDestinatarioActionPerformed() }
        jToolBar1.add(jbntDestinatario)
        jbntGrupo!!.font = Font(Font.SANS_SERIF, Font.PLAIN, 9)
        jbntGrupo!!.icon = ImageIcon(javaClass.getResource("/imagens/addusuarios.gif"))
        jbntGrupo!!.text = "Selecionar grupo"
        jbntGrupo!!.horizontalTextPosition = 0
        jbntGrupo!!.maximumSize = Dimension(90, 60)
        jbntGrupo!!.preferredSize = Dimension(90, 60)
        jbntGrupo!!.verticalTextPosition = 3
        jbntGrupo!!.addActionListener { jbntGrupoActionPerformed() }
        jToolBar1.add(jbntGrupo)
        jbntExcluir.font = Font(Font.SANS_SERIF, Font.PLAIN, 9)
        jbntExcluir.icon = ImageIcon(javaClass.getResource("/imagens/removerTodos.gif"))
        jbntExcluir.text = "Remover todos"
        jbntExcluir.horizontalTextPosition = 0
        jbntExcluir.maximumSize = Dimension(90, 60)
        jbntExcluir.verticalTextPosition = 3
        jbntExcluir.addActionListener { jbntExcluirActionPerformed() }
        jToolBar1.add(jbntExcluir)
        contentPane.add(jToolBar1, "North")
        jPanel1.layout = BorderLayout()
        jPanel2.layout = GridLayout(1, 0)
        jPanel2.border = BorderFactory.createEtchedBorder()
        jchkExportarTodos!!.font = Font("MS Sans Serif", 0, 10)
        jchkExportarTodos!!.text = "Exportar todos os destinatários"
        jchkExportarTodos!!.addActionListener { jchkExportarTodosActionPerformed() }
        jPanel2.add(jchkExportarTodos)
        jPanel1.add(jPanel2, "North")
        jScrollPane1.setViewportView(jlstDestinatarios)
        jPanel1.add(jScrollPane1, "Center")
        contentPane.add(jPanel1, "Center")
        pack()
    }

    private fun jbntExcluirActionPerformed() {
        vecDestinatario.removeAllElements()
        jlstDestinatarios!!.setListData(vecDestinatario)
        jchkExportarTodos!!.isSelected = false
        jbntDestinatario!!.isEnabled = true
        jbntGrupo!!.isEnabled = true
    }

    private fun jbntGrupoActionPerformed() {
        cursor = Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR)
        val telaPesquisarGrupo = TelaPesquisarGrupo(this, true, vecDestinatario)
        telaPesquisarGrupo.isVisible = true
        jlstDestinatarios!!.setListData(vecDestinatario)
        cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
    }

    private fun jbntDestinatarioActionPerformed() {
        cursor = Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR)
        val telaPesquisaDestinatario = TelaPesquisarDestinatario(this, true, vecDestinatario)
        telaPesquisaDestinatario.isVisible = true
        jlstDestinatarios!!.setListData(vecDestinatario)
        cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
    }

    private fun jchkExportarTodosActionPerformed() {
        cursor = Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR)
        jbntDestinatario!!.isEnabled = !jbntDestinatario!!.isEnabled
        jbntGrupo!!.isEnabled = !jbntGrupo!!.isEnabled
        vecDestinatario.removeAllElements()
        jlstDestinatarios!!.setListData(vecDestinatario)
        if (jchkExportarTodos!!.isSelected) carregaListaDestinatario()
        cursor = Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR)
    }

    private fun jbtExportarActionPerformed() {
        val geral = Geral()
        var vNomeArquivo: String
        if (vecDestinatario.size <= 0) {
            JOptionPane.showMessageDialog(
                this,
                "Não existe nenhum destinatário selecionado!",
                "Endereçador",
                JOptionPane.WARNING_MESSAGE
            )
            return
        }
        val fileChooser = JFileChooser()
        fileChooser.approveButtonText = "Salvar"
        fileChooser.dialogTitle = "Salvar arquivo"
        fileChooser.fileSelectionMode = 0
        fileChooser.isAcceptAllFileFilterUsed = true
        fileChooser.addChoosableFileFilter(FiltroArquivo("csv"))
        fileChooser.addChoosableFileFilter(FiltroArquivo("txt"))
        val result = fileChooser.showSaveDialog(this)
        if (result == 1) return
        val fileName = fileChooser.selectedFile ?: return
        var extension = Geral.getExtension(fileName)
        val imageFilter = fileChooser.fileFilter as FiltroArquivo
        vNomeArquivo = fileName.absolutePath
        if (extension == null) {
            vNomeArquivo = vNomeArquivo + "." + imageFilter.extension
            extension = imageFilter.extension
        } else if (!Geral.validaArquivo(fileName)) {
            JOptionPane.showMessageDialog(
                this,
                "Formato do arquivo inválido!\nO arquivo deve ser do tipo TXT ou CSV.",
                "Endereçador",
                JOptionPane.WARNING_MESSAGE
            )
            return
        }
        try {
            if (extension == "txt") {
                geral.exportToTxt(vNomeArquivo, vecDestinatario)
            } else {
                geral.exportToCsv(vNomeArquivo, vecDestinatario)
            }
            JOptionPane.showMessageDialog(this, "Exportação concluída!", "Endereçador", JOptionPane.WARNING_MESSAGE)
        } catch (ex: EnderecadorExcecao) {
            JOptionPane.showMessageDialog(this, ex.message, "Endereçador", JOptionPane.WARNING_MESSAGE)
        }
    }

    override fun update(o: Observable, arg: Any) {
        if (arg is DestinatarioBean) {
            val destinatario = arg
            val index = vecDestinatario.indexOf(destinatario)
            if (index != -1) {
                vecDestinatario.removeAt(index)
                vecDestinatario.add(index, destinatario)
                jlstDestinatarios!!.setListData(vecDestinatario)
            }
        } else if (arg is List<*>) {
            arg
                .filterNotNull()
                .forEach { vecDestinatario.remove(it as DestinatarioBean) }
            jlstDestinatarios!!.setListData(vecDestinatario)
        }
    }

    companion object {
        private val logger = Logger.getLogger(TelaExportarDados::class.java)
    }
}
